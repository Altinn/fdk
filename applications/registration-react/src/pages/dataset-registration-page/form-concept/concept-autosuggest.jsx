import Autosuggest from 'react-autosuggest';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import _ from 'lodash';
import { getTranslateText } from '../../../lib/translateText';
import localization from '../../../lib/localization';
import {
  extractConcepts,
  searchConcepts
} from '../../../api/search-api/concepts';

let fetchResponseThrottleTimeoutId = null;

const getSuggestionValue = suggestion =>
  getTranslateText(_.get(suggestion, 'prefLabel'));

const renderSuggestionContainer = (containerProps, children) => (
  <div {...containerProps}>
    <div className="d-flex mb-3 react_autosuggest__suggestions-heading">
      <span className="w-25 first">
        <strong>{localization.anbefaltTerm}</strong>
      </span>
      <div className="w-75 ml-5 d-flex">
        <span className="w-50">
          <strong>{localization.definition}</strong>
        </span>
        <span className="w-50 ml-5">
          <strong>{localization.responsible}</strong>
        </span>
      </div>
    </div>
    {children}
  </div>
);

const renderSuggestion = suggestion => (
  <div className="d-flex mb-3">
    <span className="w-25">
      {getTranslateText(_.get(suggestion, 'prefLabel'))}
    </span>
    <div className="w-75 ml-5 d-flex">
      <span className="w-50">
        {getTranslateText(_.get(suggestion, ['definition', 'text']))}
      </span>
      <span className="w-50 ml-5">
        {getTranslateText(
          _.get(suggestion, ['publisher', 'prefLabel']) ||
            _.get(suggestion, ['publisher', 'name'])
        )}
      </span>
    </div>
  </div>
);

export const ConceptAutosuggest = ({ addTag, ...inputProps }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [suggestions, setSuggestions] = useState([]);

  const clearSuggestions = () => setSuggestions([]);

  const fetchSuggestions = ({ value }) => {
    // Cancel the previous request
    if (fetchResponseThrottleTimeoutId) {
      clearTimeout(fetchResponseThrottleTimeoutId);
    }

    setIsLoading(true);

    searchConcepts({
      prefLabel: value,
      returnfields: 'uri,definition.text,publisher.prefLabel,publisher.name'
    })
      .then(extractConcepts)
      .then(concepts => {
        fetchResponseThrottleTimeoutId = setTimeout(() => {
          setIsLoading(false);
          setSuggestions(concepts);
        }, 250);
      })
      .catch(console.error);
  };

  return (
    <>
      <Autosuggest
        suggestions={suggestions}
        shouldRenderSuggestions={value => value && value.trim().length > 0}
        getSuggestionValue={suggestion => getSuggestionValue(suggestion)}
        renderSuggestion={renderSuggestion}
        renderSuggestionsContainer={({ containerProps, children }) =>
          renderSuggestionContainer(containerProps, children)
        }
        inputProps={inputProps}
        onSuggestionSelected={(e, { suggestion }) => {
          addTag(suggestion);
        }}
        onSuggestionsClearRequested={clearSuggestions}
        onSuggestionsFetchRequested={fetchSuggestions}
      />
      {isLoading && (
        <i
          className="fa fa-spinner fa-spin"
          style={{ position: 'absolute', right: '10px', top: '12px' }}
        />
      )}
    </>
  );
};
ConceptAutosuggest.propTypes = {
  addTag: PropTypes.func.isRequired
};
