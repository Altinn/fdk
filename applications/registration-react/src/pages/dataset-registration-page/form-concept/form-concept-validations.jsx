import { validateMinTwoChars } from '../../../validation/validation';
import localization from '../../../lib/localization';

const validate = values => {
  let errors = {};
  const { keyword } = values;

  if (keyword) {
    keyword.forEach(item => {
      errors = validateMinTwoChars(
        'keyword',
        item[localization.getLanguage()],
        errors
      );
    });
  }
  return errors;
};

export default validate;
