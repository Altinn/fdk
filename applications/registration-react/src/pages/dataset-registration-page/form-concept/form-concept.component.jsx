import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import _ from 'lodash';

import localization from '../../../lib/localization';
import Helptext from '../../../components/helptext/helptext.component';
import InputTagsFieldConcepts from './input-tags-concepts/input-tags-concepts.component';
import InputTagsFieldArray from '../../../components/field-input-tags-objects/field-input-tags-objects.component';
import MultilingualField from '../../../components/multilingual-field/multilingual-field.component';

export const FormConcept = ({ syncErrors, languages }) => {
  return (
    <form>
      <div className="form-group">
        <Helptext
          title={localization.schema.concept.helptext.content}
          term="Dataset_content"
        />
        <Field
          name="concepts"
          type="text"
          component={InputTagsFieldConcepts}
          label={localization.schema.concept.conceptLabel}
          fieldLabel="no"
        />
      </div>
      <div className="form-group">
        <Helptext
          title={localization.schema.concept.helptext.keyword}
          term="Dataset_keyword"
        />
        <MultilingualField
          name="keyword"
          type="text"
          languages={languages}
          component={InputTagsFieldArray}
          label={localization.schema.concept.keywordLabel}
          // fieldLabel={localization.getLanguage()}
        />
        {_.get(syncErrors, 'keyword') && (
          <div className="alert alert-danger mt-3">
            {_.get(syncErrors, 'keyword')[localization.getLanguage()]}
          </div>
        )}
      </div>
    </form>
  );
};

FormConcept.defaultProps = {
  syncErrors: null,
  languages: []
};
FormConcept.propTypes = {
  syncErrors: PropTypes.object,
  languages: PropTypes.array
};
