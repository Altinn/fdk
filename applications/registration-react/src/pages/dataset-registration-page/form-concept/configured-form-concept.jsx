import { reduxForm } from 'redux-form';

import { FormConcept } from './form-concept.component';
import validate from './form-concept-validations';
import { asyncValidateDatasetInvokePatch } from '../formsLib/asyncValidateDatasetInvokePatch';

const config = {
  form: 'concept',
  validate,
  asyncValidateDatasetInvokePatch
};

export const ConfiguredFormConcept = reduxForm(config)(FormConcept);
