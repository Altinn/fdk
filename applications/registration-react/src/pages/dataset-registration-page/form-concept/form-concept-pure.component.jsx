import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';

import localization from '../../../services/localization';
import { Helptext } from '../../../components/helptext/helptext.component';
import LookupTagsInputField from '../../../components/fields/lookup-tags-input-field/lookup-tags-input-field.component';
import { ConceptTagReadOnlyField } from './concept-tags-readonly-field/concept-tags-readonly-field';
import TagsInputFieldArray from '../../../components/fields/field-input-tags-objects/tags-input-field-array.component';
import MultilingualField from '../../../components/multilingual-field/multilingual-field.component';
import { getTranslateText } from '../../../services/translateText';
import { renderConceptAutosuggestForTagsInput } from './concept-autosuggest';
import InputFieldReadonly from '../../../components/fields/field-input-readonly/field-input-readonly.component';

export const FormConceptPure = ({ languages, errors, isReadOnly }) => {
  return (
    <form>
      <div className="form-group">
        <Helptext
          title={localization.schema.concept.helptext.content}
          term="Dataset_content"
        />
        {isReadOnly && (
          <Field
            name="concepts"
            component={ConceptTagReadOnlyField}
            languages={languages}
          />
        )}
        {!isReadOnly && (
          <Field
        name="concepts"
        component={LookupTagsInputField}
        getTagFromItem={item => getTranslateText(item.prefLabel)}
        renderLookupInput={renderConceptAutosuggestForTagsInput}
      />
        )}
      </div>
      <div className="form-group">
        <Helptext
          title={localization.schema.concept.helptext.keyword}
          term="Dataset_keyword"
        />
        {isReadOnly && (
          <MultilingualField
            name="keyword"
            languages={languages}
            component={InputFieldReadonly}
            label={localization.schema.concept.keywordLabel}
          />
        )}
        {!isReadOnly && (
          <MultilingualField
            name="keyword"
            languages={languages}
            component={TagsInputFieldArray}
            label={localization.schema.concept.keywordLabel}
          />
        )}

        {errors && errors.keyword && (
          <div className="alert alert-danger mt-3">
            {errors.keyword[localization.getLanguage()]}
          </div>
        )}
      </div>
    </form>
  );
};

FormConceptPure.defaultProps = {
  errors: {},
  languages: [],
  isReadOnly: false
};
FormConceptPure.propTypes = {
  errors: PropTypes.object,
  languages: PropTypes.array,
  isReadOnly: PropTypes.bool
};
