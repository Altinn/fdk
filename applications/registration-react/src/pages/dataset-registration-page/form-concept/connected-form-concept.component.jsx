import { getFormSyncErrors } from 'redux-form';
import { connect } from 'react-redux';
import _ from 'lodash';

import { ConfiguredFormConcept } from './configured-form-concept';

const mapStateToProps = (state, ownProps) => {
  const { datasetItem } = ownProps;
  return {
    initialValues: {
      concepts:
        _.get(datasetItem, 'concepts', []).length > 0
          ? _.get(datasetItem, 'concepts')
          : [],
      keyword:
        _.get(datasetItem, 'keyword', []).length > 0
          ? _.get(datasetItem, 'keyword')
          : []
    },
    syncErrors: getFormSyncErrors('concept')(state)
  };
};

export const ConnectedFormConcept = connect(mapStateToProps)(
  ConfiguredFormConcept
);
