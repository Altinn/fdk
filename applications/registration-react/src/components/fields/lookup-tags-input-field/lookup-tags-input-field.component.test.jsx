import React from 'react';
import { shallow } from 'enzyme';
import LookupTagsInputField from './lookup-tags-input-field.component';
import { renderConceptAutosuggestForTagsInput } from '../../../pages/dataset-registration-page/form-concept/concept-autosuggest';

let defaultProps;
let wrapper;

beforeEach(() => {
  defaultProps = {
    input: {
      name: 'subject',
      value: [
        {
          uri: 'https://data-david.github.io/Begrep/begrep/Enhet',
          identifier: 'https://data-david.github.io/Begrep/begrep/Enhet',
          prefLabel: { no: 'enhet' },
          altLabel: null,
          definition: {
            no: 'alt som er registrert med et organisasjonsnummer '
          },
          note: {
            no:
              'Alle hovedenheter, underenheter og organisasjonsledd som er identifisert med et organisasjonsnummer.'
          },
          source: 'https://jira.brreg.no/browse/BEGREP-208',
          creator: {
            uri: 'http://data.brreg.no/enhetsregisteret/enhet/974760673'
          },
          inScheme: ['http://data-david.github.io/vokabular/Befolkning'],
          datasets: null
        }
      ]
    },
    renderLookupInput: renderConceptAutosuggestForTagsInput
  };
  wrapper = shallow(<LookupTagsInputField {...defaultProps} />);
});

test('should render LookupTagsInputField correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
