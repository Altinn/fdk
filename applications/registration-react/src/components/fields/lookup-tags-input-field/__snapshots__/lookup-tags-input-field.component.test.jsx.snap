// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LookupTagsInputField correctly 1`] = `
<div
  className="pl-2"
>
  <div
    className="d-flex align-items-center"
  >
    <TagsInput
      addKeys={
        Array [
          9,
          13,
        ]
      }
      addOnBlur={false}
      addOnPaste={false}
      className="fdk-reg-input-tags"
      disabled={false}
      focusedClassName="react-tagsinput--focused"
      inputProps={
        Object {
          "placeholder": "",
        }
      }
      maxTags={-1}
      onChange={[Function]}
      onlyUnique={false}
      pasteSplit={[Function]}
      preventSubmit={true}
      removeKeys={
        Array [
          8,
        ]
      }
      renderInput={[Function]}
      renderLayout={[Function]}
      renderTag={[Function]}
      tagDisplayProp={null}
      tagProps={
        Object {
          "className": "react-tagsinput-tag",
          "classNameRemove": "react-tagsinput-remove",
        }
      }
      validationRegex={/\\.\\*/}
      value={
        Array [
          Object {
            "altLabel": null,
            "creator": Object {
              "uri": "http://data.brreg.no/enhetsregisteret/enhet/974760673",
            },
            "datasets": null,
            "definition": Object {
              "no": "alt som er registrert med et organisasjonsnummer ",
            },
            "identifier": "https://data-david.github.io/Begrep/begrep/Enhet",
            "inScheme": Array [
              "http://data-david.github.io/vokabular/Befolkning",
            ],
            "note": Object {
              "no": "Alle hovedenheter, underenheter og organisasjonsledd som er identifisert med et organisasjonsnummer.",
            },
            "prefLabel": Object {
              "no": "enhet",
            },
            "source": "https://jira.brreg.no/browse/BEGREP-208",
            "uri": "https://data-david.github.io/Begrep/begrep/Enhet",
          },
        ]
      }
    />
  </div>
</div>
`;
